x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/HttpUtils.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.inference.external.http;
+
+import org.apache.http.client.methods.HttpRequestBase;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.xpack.inference.logging.ThrottlerManager;
+
+import static org.elasticsearch.core.Strings.format;
+
+public class HttpUtils {
+
+    public static void checkForFailureStatusCode(
+        ThrottlerManager throttlerManager,
+        Logger logger,
+        HttpRequestBase request,
+        HttpResult result
+    ) {
+        if (result.response().getStatusLine().getStatusCode() >= 300) {
```

Comment:
> >= 300 rather than >=400, is this because the client does not handle redirection.

Correct, the default redirection strategy for the apache client does not follow redirects for put and post and I haven't added code to implement it yet.

https://hc.apache.org/httpcomponents-client-4.5.x/current/httpclient/apidocs/

Doesn't look like it'd be hard to do based on this: https://stackoverflow.com/questions/5169468/handling-httpclient-redirects

I'll add it to the list of TODOs

> Can you log the >= 300, < 400 response codes as a separate message - "Unhandled redirection for request [%s] status [%s]"

Yep will do 👍